pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -551,29 +624,46 @@ protected void handleConnect(CommandConnect connect) {
             if (sslHandler != null) {
                 sslSession = ((SslHandler) sslHandler).engine().getSession();
             }
-            originalPrincipal = getOriginalPrincipal(
-                connect.hasOriginalAuthData() ? connect.getOriginalAuthData() : null,
-                connect.hasOriginalAuthMethod() ? connect.getOriginalAuthMethod() : null,
-                connect.hasOriginalPrincipal() ? connect.getOriginalPrincipal() : null,
-                sslSession);
 
             authState = authenticationProvider.newAuthState(clientData, remoteAddress, sslSession);
             authenticationData = authState.getAuthDataSource();
-            doAuthentication(clientData, clientProtocolVersion, clientVersion);
+            state = doAuthentication(clientData, clientProtocolVersion, clientVersion);
```

Comment:
not for this patch, but it seems strange that this is a synchronous call, given that authentication could be calling out to a remote system.

Reply:
Yes, initially all auth methods were completely stateless and therefore didn't require any asynchronousness in the provider API. This is not true now in general and we'll have to make a bit of a refactor to achieve that.